compiler/sem/semexprs.nim

```diff
@@ -2141,55 +2137,78 @@ proc semProcBody(c: PContext, n: PNode): PNode =
   closeScope(c)
 
 proc semYieldVarResult(c: PContext, n: PNode, restype: PType): PNode =
+  ## Validates that the yield's expression is a tuple constructor when `restype`
+  ## is a tuple type that contains views. Also introduces ``nkHiddenAddr``
+  ## where needed
   addInNimDebugUtils(c.config, "semYieldVarResult", n, result)
-  result = n
+  result = copyNode(n)
+
   let t = skipTypes(restype, {tyGenericInst, tyAlias, tySink})
+  var hasError = false
+
   case t.kind
   of tyVar, tyLent:
     let unwrappedValue =
-      case result[0].kind
+      case n[0].kind
       of nkHiddenStdConv, nkHiddenSubConv:
-        result[0][1]
+        n[0][1]
       else:
-        result[0]
-    result[0] = takeImplicitAddr(c, unwrappedValue, t.kind == tyLent)
+        n[0]
+
+    result.add takeImplicitAddr(c, t, unwrappedValue)
+    hasError = result[0].isError
   of tyTuple:
-    var illformedAst = false
-      ## in some cases syntax issues that we can diagnose precisely enough
-      ## occur, this happens often with tuples (parentheses) so we use this to
-      ## flag if we end up in such a scenario
+    var
+      notTupleConstr = false
+        # if the tuple contains a view (i.e. ``var|lent``), the yield
+        # expression *must* be a tuple constructor
+        # XXX: this doesn't seem right in the presence of first-class view
```

Comment:
Same as the previous: something that needs to be fixed or just a general observation of the weirdness of this code? Or maybe it is a QUESTION about possible suggestions for refactoring (again, note for latter readers)